### PR TITLE
perf: Optimize prefix-group processing in `PartialSortExec`

### PR DESCRIPTION
**Pull request by lyne7-sc**
## Which issue does this PR close?

- Closes #.

## Rationale for this change

`PartialSortStream` concatenates each incoming batch with the buffered incomplete prefix before looking for a completed prefix boundary.

When one prefix spans multiple batches, this repeatedly copies rows accumulated from earlier batches. When a batch contains multiple completed prefix groups, the completed region is also sorted using the full ordering even though the prefix ordering is already satisfied.

This adds unnecessary copying and sorting work, especially for wide batches and workloads containing multiple prefix groups per batch.

## What changes are included in this PR?

- Keep the unfinished trailing prefix as a list of `RecordBatch` slices instead of concatenating it with every incoming batch.
- Detect completed prefix groups within each batch and compare prefix expression results directly across batch boundaries.
- Sort completed groups only by the remaining suffix ordering.
- Concatenate a single completed prefix at most once.
- For multiple completed prefixes, evaluate suffix expressions once per source batch and materialize the final output with one interleave operation.
- Preserve fetch handling, input release, empty-batch handling, and zero-column batch semantics.

## What is the testing strategy for this PR?

The implementation is covered by the existing `PartialSortExec` unit tests and the additional SQL logic test cases in `group_by.slt`.

## Are there any user-facing changes?

No. This is an internal performance improvement.

## Benchmark
```
group                                 main                                   optimized
-----                                 ----                                   ---------
partial_sort/rows_per_prefix/100      3.10     12.1±0.13ms        ? ?/sec    1.00      3.9±0.05ms        ? ?/sec
partial_sort/rows_per_prefix/1000     4.17     13.2±0.07ms        ? ?/sec    1.00      3.2±0.14ms        ? ?/sec
partial_sort/rows_per_prefix/10000    1.30      2.8±0.02ms        ? ?/sec    1.00      2.1±0.07ms        ? ?/sec
partial_sort/rows_per_prefix/20000    1.41      3.0±0.22ms        ? ?/sec    1.00      2.1±0.15ms        ? ?/sec
partial_sort/rows_per_prefix/5000     3.85     12.4±0.27ms        ? ?/sec    1.00      3.2±0.03ms        ? ?/sec
partial_sort/rows_per_prefix/8192     1.56      2.8±0.05ms        ? ?/sec    1.00  1780.8±34.70µs        ? ?/sec
```

### CODE REVIEW

**Review by ryux1 — APPROVED** [review]
Traced the stream state across continued prefixes, batch-boundary changes, multiple completed prefixes, end-of-input, and fetch exhaustion. The logical prefix offsets and batch-local fragment mapping are consistent; first-occurrence prefix order is preserved, suffix ordering is applied independently per completed group, and fetch is consumed across groups before the next poll closes/releases the input. The explicit zero-column materialization also preserves row counts. The added SQL cases cover the important cross-batch paths, and the reported CI matrix is green.

File: datafusion/physical-plan/src/sorts/partial_sort.rs [hunk 1]
```diff
@@ -583,50 +798,142 @@ impl PartialSortStream {
...
+    }
+
+    fn get_prefix_ranges(&self, batch: &RecordBatch) -> Result<Vec<Range<usize>>> {
+        let common_prefix_sort_keys = (0..self.common_prefix_length)
+            .map(|idx| self.expr[idx].evaluate_to_sort_column(batch))
+            .collect::<Result<Vec<_>>>()?;
+        evaluate_partition_ranges(batch.num_rows(), &common_prefix_sort_keys)
+    }
+
+    fn prefix_changed_at_batch_boundary(
```
Comment (kosiew) [inline datafusion/physical-plan/src/sorts/partial_sort.rs:914]: Could we add a cross-batch regression test where the prefix key contains repeated NULL and NaN values, including `nulls_first` and descending sort options? It would be useful to compare the result against `SortExec` as well. `prefix_changed_at_batch_boundary` now determines boundary equality independently using `make_comparator`, while the in-batch groups come from `evaluate_partition_ranges`, so a test like this would help make sure both paths stay consistent for these special values.

**Review by kosiew — APPROVED** [review]
@lyne7-sc,

Thanks for working on this. The new prefix-group processing looks like a solid improvement, especially avoiding repeated concatenation for prefixes that span batches and limiting sorting work to the suffix keys.

I have one non-blocking suggestion around coverage for special prefix values. Otherwise, this looks good to me.

File: datafusion/physical-plan/src/sorts/partial_sort.rs (hunk 1 above)
Comment (lyne7-sc) [reply datafusion/physical-plan/src/sorts/partial_sort.rs:914]: Thanks for the suggestion! Added a cross-batch test with repeated `NULL` and `NaN` prefixes, covering ascending/descending order and `NULLS FIRST/LAST`, and comparing the results against `SortExec`.

**Comment by kosiew** [comment]
🚀 
@lyne7-sc 
Thank you for your contribution.
